cpp/src/io/parquet/page_hdr.cu

```diff
@@ -121,70 +122,150 @@ __device__ void skip_struct_field(byte_stream_s *bs, int t)
   } while (rep_cnt || struct_depth);
 }
 
-#define PARQUET_BEGIN_STRUCT(fn)         \
-  __device__ bool fn(byte_stream_s *bs)  \
-  {                                      \
-    int fld = 0;                         \
-    for (;;) {                           \
-      int c, t, f;                       \
-      c = getb(bs);                      \
-      if (!c) break;                     \
-      f   = c >> 4;                      \
-      t   = c & 0xf;                     \
-      fld = (f) ? fld + f : get_i32(bs); \
-      switch (fld) {
-#define PARQUET_FLD_ENUM(id, m, mt)    \
-  case id:                             \
-    bs->m = (mt)get_i32(bs);           \
-    if (t != ST_FLD_I32) return false; \
-    break;
-
-#define PARQUET_FLD_INT32(id, m)       \
-  case id:                             \
-    bs->m = get_i32(bs);               \
-    if (t != ST_FLD_I32) return false; \
-    break;
-
-#define PARQUET_FLD_STRUCT(id, m)                   \
-  case id:                                          \
-    if (t != ST_FLD_STRUCT || !m(bs)) return false; \
-    break;
-
-#define PARQUET_END_STRUCT()                \
-  default: skip_struct_field(bs, t); break; \
-    }                                       \
-    }                                       \
-    return true;                            \
+struct ParquetFieldInt32 {
```

Comment:
We need some docs for the new code, this code is not trivial in any way. Simple comment about what the function does + more detailed explanation of the more complex functions.